### PR TITLE
Feature/ksk views

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Description

Create public APIs for ksk views, for compiler CUDA use case, additional engine + FFI to follow, and then fixture to improve delivery time

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [ ] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [ ] The draft release description has been updated
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_creation.rs [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+    /// use concrete_core::prelude::*;
+    /// # use std::error::Error;
+    ///
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// // DISCLAIMER: the parameters used here are only for test purpose, and are not secure.
+    /// let input_lwe_dimension = LweDimension(6);
+    /// let output_lwe_dimension = LweDimension(3);
+    /// let decomposition_level_count = DecompositionLevelCount(2);
+    /// let decomposition_base_log = DecompositionBaseLog(8);
+    /// let noise = Variance(2_f64.powf(-25.));
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_creation.rs:110]: -50

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_consuming_retrieval.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    /// use concrete_core::prelude::*;
+    /// # use std::error::Error;
+    ///
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// // DISCLAIMER: the parameters used here are only for test purpose, and are not secure.
+    /// let input_lwe_dimension = LweDimension(6);
+    /// let output_lwe_dimension = LweDimension(3);
+    /// let decomposition_level_count = DecompositionLevelCount(2);
+    /// let decomposition_base_log = DecompositionBaseLog(8);
+    /// let noise = Variance(2_f64.powf(-25.));
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_consuming_retrieval.rs:78]: -50

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_consuming_retrieval.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_consuming_retrieval.rs:195]: -50

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_consuming_retrieval.rs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_consuming_retrieval.rs:310]: -50

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_creation.rs [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_creation.rs:291]: -50

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_creation.rs [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_keyswitch_key_creation.rs:472]: -50

File: concrete-core/src/specification/engines/lwe_keyswitch_key_creation.rs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        decomposition_level_count: DecompositionLevelCount,
+        ciphertext_modulus_log: usize,
+    ) -> Result<(), Self> {
+        if decomposition_base_log.0 == 0 {
+            return Err(Self::NullDecompositionBaseLog);
+        }
+        if decomposition_level_count.0 == 0 {
+            return Err(Self::NullDecompositionLevelCount);
+        }
+        if decomposition_base_log.0 * decomposition_level_count.0 > ciphertext_modulus_log {
```
Comment (agnesLeroy) [inline concrete-core/src/specification/engines/lwe_keyswitch_key_creation.rs:31]: Is this check correct? It seems to me it should be:
```
log_2(base_log) * l < ciphertext_modulus_log

**Review by agnesLeroy — COMMENTED** [review]
Hello! Thank you very much for this PR. Are you planning to add serialization in another PR?

**Comment by IceTDrinker** [comment]
KSK serialization already exists, I won't add it for the mut view because it's just a hack for the compiler to be able to manipulate keys for the GPU

File: concrete-core/src/specification/engines/lwe_keyswitch_key_creation.rs (hunk 7 above)
Comment (IceTDrinker) [reply concrete-core/src/specification/engines/lwe_keyswitch_key_creation.rs:31]: yes, base_log is the power of 2 we use/a number of bits for a given base, we want the base * number of levels to be smaller than the number of bits we can represent

**Comment by IceTDrinker** [comment]
waiting for tests to complete to push updated branch

**Review by agnesLeroy — APPROVED** [review]

**Pull request merged**